File: fsharp/07/IntComputer.fs
module IntComputer

open System.IO

type ParameterMode =
    | Address
    | Immediate
    | Relative

type ComputerState =
    { Memory: int64 array
      InstructionPointer: int
      RelativeBase: int
      FullStop: bool }

let ParseProgram path = File.ReadAllText(path).Split(",") |> Array.map int64

let ReadParameterMode address position (program: int64 array) =
    let mode = (program.[address] / ((pown 10 (2 + position)) |> int64)) % 10L
    match mode with
    | 0L -> Address
    | 1L -> Immediate
    | 2L -> Relative

let ReadParameter (program: int64 array) address relativeBase mode =
    match mode with
    | Address -> program.[address |> int]
    | Immediate -> address
    | Relative -> program.[(relativeBase + (address |> int))]

let GetWritePosition address relativeBase (mode: ParameterMode) =
    match mode with
    | Address -> address |> int
    | Relative -> relativeBase + (address |> int)

let rec ExecuteProgram instructionPointer relativeBase reader writer earlyStopper (program: int64 array) =
    let instruction = program.[instructionPointer] % 100L

    let newPointer =
        match instruction with
        | 1L ->
            let a =
                program
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase
            let b =
                program
                |> ReadParameterMode instructionPointer 1
                |> ReadParameter program program.[instructionPointer + 2] relativeBase

            let p =
                program
                |> ReadParameterMode instructionPointer 2
                |> GetWritePosition program.[instructionPointer + 3] relativeBase
            program.[p] <- a + b

            Some(instructionPointer + 4), None
        | 2L ->
            let a =
                program
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase
            let b =
                program
                |> ReadParameterMode instructionPointer 1
                |> ReadParameter program program.[instructionPointer + 2] relativeBase

            let p =
                program
                |> ReadParameterMode instructionPointer 2
                |> GetWritePosition program.[instructionPointer + 3] relativeBase
            program.[p] <- a * b

            Some(instructionPointer + 4), None
        | 3L ->
            let p =
                program
                |> ReadParameterMode instructionPointer 0
                |> GetWritePosition program.[instructionPointer + 1] relativeBase
            program.[p] <- reader()

            Some(instructionPointer + 2), None
        | 4L ->
            let a =
                program
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase
            writer (a)

            Some(instructionPointer + 2), None
        | 5L ->
            let value =
                program 
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase
            let target =
                program
                |> ReadParameterMode instructionPointer 1
                |> ReadParameter program program.[instructionPointer + 2] relativeBase

            match value with
            | 0L -> Some(instructionPointer + 3), None
            | _ -> Some(target |> int), None
        | 6L ->
            let value =
                program
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase
            let target =
                program
                |> ReadParameterMode instructionPointer 1
                |> ReadParameter program program.[instructionPointer + 2] relativeBase

            match value with
            | 0L -> Some(target |> int), None
            | _ -> Some(instructionPointer + 3), None
        | 7L ->
            let a =
                program
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase
            let b =
                program
                |> ReadParameterMode instructionPointer 1
                |> ReadParameter program program.[instructionPointer + 2] relativeBase

            let p =
                program
                |> ReadParameterMode instructionPointer 2
                |> GetWritePosition program.[instructionPointer + 3] relativeBase
            program.[p] <- if a < b then 1L else 0L

            Some(instructionPointer + 4), None
        | 8L ->
            let a =
                program
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase
            let b =
                program
                |> ReadParameterMode instructionPointer 1
                |> ReadParameter program program.[instructionPointer + 2] relativeBase

            let p =
                program
                |> ReadParameterMode instructionPointer 2
                |> GetWritePosition program.[instructionPointer + 3] relativeBase
            program.[p] <- if a = b then 1L else 0L

            Some(instructionPointer + 4), None
        | 9L ->
            let a =
                program
                |> ReadParameterMode instructionPointer 0
                |> ReadParameter program program.[instructionPointer + 1] relativeBase

            Some(instructionPointer + 2), Some(a |> int)
        | 99L -> None, None


    let requiresEarlyStop =
        match earlyStopper with 
        | Some stopFunction ->
            instruction |> stopFunction
        | None ->
            false

    match newPointer with
    | Some ip, Some b ->
        match requiresEarlyStop with
        | true ->
            { ComputerState.Memory = program
              ComputerState.InstructionPointer = ip
              ComputerState.RelativeBase = b
              FullStop = false }
        | false ->
            program |> ExecuteProgram ip b reader writer earlyStopper
    | Some ip, None -> 
        match requiresEarlyStop with
        | true ->
            { ComputerState.Memory = program
              ComputerState.InstructionPointer = ip
              ComputerState.RelativeBase = relativeBase
              FullStop = false }
        | false ->
            program |> ExecuteProgram ip relativeBase reader writer earlyStopper
    | None, _ ->
        { ComputerState.Memory = program
          ComputerState.InstructionPointer = instructionPointer
          ComputerState.RelativeBase = relativeBase
          ComputerState.FullStop = true }
 

File: csharp/07/IntComputer.cs
using System;
using System.Collections.Concurrent;

namespace _07
{
    public sealed class IntComputer
    {
        public IntComputer()
        {
            IO = new InputOutput();
        }

        public IntComputer(BlockingCollection<int> input, BlockingCollection<int> output)
        {
            IO = new InputOutput(input, output);
        }

        public InputOutput IO { get; }

        public void ExecuteProgram(int[] program)
        {
            var pointer = 0;

            while (ExecuteInstruction(program, ref pointer))
            {
            }
        }

        private bool ExecuteInstruction(int[] program, ref int pointer)
        {
            var instruction = program[pointer];
            var operation = (Operation)(instruction % 100);

            switch (operation)
            {
                case Operation.Add:
                    Add(program, ref pointer, instruction);
                    break;
                case Operation.Multiply:
                    Multiply(program, ref pointer, instruction);
                    break;
                case Operation.Read:
                    Read(program, ref pointer, IO);
                    break;
                case Operation.Write:
                    Write(program, ref pointer, instruction, IO);
                    break;
                case Operation.JFT:
                    JumpIfTrue(program, ref pointer, instruction);
                    break;
                case Operation.JFF:
                    JumpIfFalse(program, ref pointer, instruction);
                    break;
                case Operation.LessThan:
                    LessThan(program, ref pointer, instruction);
                    break;
                case Operation.Equals:
                    IsEqual(program, ref pointer, instruction);
                    break;
                case Operation.Terminate:
                    return false;
            }

            return true;
        }

        private static void IsEqual(int[] program, ref int pointer, int instruction)
        {
            var a = ReadParameter(program, program[pointer + 1], ReadParameterMode(instruction, 0));
            var b = ReadParameter(program, program[pointer + 2], ReadParameterMode(instruction, 1));

            program[ReadParameter(program, program[pointer + 3], ParameterMode.Immediate)] = a == b ? 1 : 0;

            pointer += 4;
        }

        private static void LessThan(int[] program, ref int pointer, int instruction)
        {
            var a = ReadParameter(program, program[pointer + 1], ReadParameterMode(instruction, 0));
            var b = ReadParameter(program, program[pointer + 2], ReadParameterMode(instruction, 1));

            program[ReadParameter(program, program[pointer + 3], ParameterMode.Immediate)] = a < b ? 1 : 0;

            pointer += 4;
        }

        private static void JumpIfFalse(int[] program, ref int pointer, int instruction)
        {
            var valueToTest = ReadParameter(program, program[pointer + 1], ReadParameterMode(instruction, 0));
            var target = ReadParameter(program, program[pointer + 2], ReadParameterMode(instruction, 1));

            pointer = valueToTest == 0 ? target : pointer + 3;
        }

        private static void JumpIfTrue(int[] program, ref int pointer, int instruction)
        {
            var valueToTest = ReadParameter(program, program[pointer + 1], ReadParameterMode(instruction, 0));
            var target = ReadParameter(program, program[pointer + 2], ReadParameterMode(instruction, 1));

            pointer = valueToTest != 0 ? target : pointer + 3;
        }

        private static void Write(int[] program, ref int pointer, int instruction, InputOutput io)
        {
            var value = ReadParameter(program, program[pointer + 1], ReadParameterMode(instruction, 0));

            io.Write(value);

            pointer += 2;
        }

        private static void Read(int[] program, ref int pointer, InputOutput io)
        {
            program[ReadParameter(program, program[pointer + 1], ParameterMode.Immediate)] = io.Read();
            pointer += 2;
        }

        private static void Multiply(int[] program, ref int pointer, int instruction)
        {
            var a = ReadParameter(program, program[pointer + 1], ReadParameterMode(instruction, 0));
            var b = ReadParameter(program, program[pointer + 2], ReadParameterMode(instruction, 1));

            program[ReadParameter(program, program[pointer + 3], ParameterMode.Immediate)] = a * b;
            pointer += 4;
        }

        private static void Add(int[] program, ref int pointer, int instruction)
        {
            var a = ReadParameter(program, program[pointer + 1], ReadParameterMode(instruction, 0));
            var b = ReadParameter(program, program[pointer + 2], ReadParameterMode(instruction, 1));

            program[ReadParameter(program, program[pointer + 3], ParameterMode.Immediate)] = a + b;
            pointer += 4;
        }

        private static ParameterMode ReadParameterMode(int instruction, int position)
        {
            instruction /= (int)Math.Pow(10, 2 + position);
            instruction %= 10;

            return (ParameterMode)instruction;
        }

        private static int ReadParameter(int[] program, int address, ParameterMode mode)
        {
            return mode switch
            {
                ParameterMode.Address => program[address],
                ParameterMode.Immediate => address,
                _ => throw new Exception()
            };
        }

        public class InputOutput
        {
            // underlying concurrent queue will ensure a FIFO execution order
            public BlockingCollection<int> Input { get; }
            public BlockingCollection<int> Output { get; }

            public InputOutput()
            {
                Input = new BlockingCollection<int>(new ConcurrentQueue<int>());
                Output = new BlockingCollection<int>(new ConcurrentQueue<int>());
            }

            public InputOutput(BlockingCollection<int> input, BlockingCollection<int> output)
            {
                Input = input;
                Output = output;
            }

            public void AddInput(int value)
            {
                Input.Add(value);
            }

            public int Read()
            {
                Input.TryTake(out var value, TimeSpan.FromMilliseconds(100));

                return value;
            }

            public int ReadOutput()
            {
                Output.TryTake(out var value, TimeSpan.FromMilliseconds(100));

                return value;
            }

            public void Write(int value)
            {
                Output.Add(value);
            }
        }

        private enum Operation
        {
            Add = 1,
            Multiply = 2,
            Read = 3,
            Write = 4,
            JFT = 5,
            JFF = 6,
            LessThan = 7,
            Equals = 8,
            Terminate = 99
        }

        private enum ParameterMode
        {
            Address = 0,
            Immediate = 1
        }
    }
}
